File: Server/src/Server/Network.java
package Server;

import java.io.*;
import java.net.*;
import java.util.HashMap;

public class Network {
	private Socket client = null;
	private HashMap<Character, String> thaiCharKey;
	public Network(Socket client) {
		this.client = client;
		this.thaiCharKey = new HashMap<Character, String>();
        this.thaiCharKey.put('¡', "'161'");
        this.thaiCharKey.put('¢', "'162'");
        this.thaiCharKey.put('£', "'163'");
        this.thaiCharKey.put('¤', "'164'");
        this.thaiCharKey.put('¥', "'165'");
        this.thaiCharKey.put('¦', "'166'");
        this.thaiCharKey.put('§', "'167'");
        this.thaiCharKey.put('¨', "'168'");
        this.thaiCharKey.put('©', "'169'");
        this.thaiCharKey.put('ª', "'170'");
        this.thaiCharKey.put('«', "'171'");
        this.thaiCharKey.put('¬', "'172'");
        this.thaiCharKey.put('­', "'173'");
        this.thaiCharKey.put('®', "'174'");
        this.thaiCharKey.put('¯', "'175'");
        this.thaiCharKey.put('°', "'176'");
        this.thaiCharKey.put('±', "'177'");
        this.thaiCharKey.put('²', "'178'");
        this.thaiCharKey.put('³', "'179'");
        this.thaiCharKey.put('´', "'180'");
        this.thaiCharKey.put('µ', "'181'");
        this.thaiCharKey.put('¶', "'182'");
        this.thaiCharKey.put('·', "'183'");
        this.thaiCharKey.put('¸', "'184'");
        this.thaiCharKey.put('¹', "'185'");
        this.thaiCharKey.put('º', "'186'");
        this.thaiCharKey.put('»', "'187'");
        this.thaiCharKey.put('¼', "'188'");
        this.thaiCharKey.put('½', "'189'");
        this.thaiCharKey.put('¾', "'190'");
        this.thaiCharKey.put('¿', "'191'");
        this.thaiCharKey.put('À', "'192'");
        this.thaiCharKey.put('Á', "'193'");
        this.thaiCharKey.put('Â', "'194'");
        this.thaiCharKey.put('Ã', "'195'");
        this.thaiCharKey.put('Ä', "'196'");
        this.thaiCharKey.put('Å', "'197'");
        this.thaiCharKey.put('Æ', "'198'");
        this.thaiCharKey.put('Ç', "'199'");
        this.thaiCharKey.put('È', "'200'");
        this.thaiCharKey.put('É', "'201'");
        this.thaiCharKey.put('Ê', "'202'");
        this.thaiCharKey.put('Ë', "'203'");
        this.thaiCharKey.put('Ì', "'204'");
        this.thaiCharKey.put('Í', "'205'");
        this.thaiCharKey.put('Î', "'206'");
        this.thaiCharKey.put('Ï', "'207'");
        this.thaiCharKey.put('Ð', "'208'");
        this.thaiCharKey.put('Ñ', "'209'");
        this.thaiCharKey.put('Ò', "'210'");
        this.thaiCharKey.put('Ó', "'211'");
        this.thaiCharKey.put('Ô', "'212'");
        this.thaiCharKey.put('Õ', "'213'");
        this.thaiCharKey.put('Ö', "'214'");
        this.thaiCharKey.put('×', "'215'");
        this.thaiCharKey.put('Ø', "'216'");
        this.thaiCharKey.put('Ù', "'217'");
        this.thaiCharKey.put('Ú', "'218'");
        this.thaiCharKey.put('ß', "'223'");
        this.thaiCharKey.put('à', "'224'");
        this.thaiCharKey.put('á', "'225'");
        this.thaiCharKey.put('â', "'226'");
        this.thaiCharKey.put('ã', "'227'");
        this.thaiCharKey.put('ä', "'228'");
        this.thaiCharKey.put('å', "'229'");
        this.thaiCharKey.put('æ', "'230'");
        this.thaiCharKey.put('ç', "'231'");
        this.thaiCharKey.put('è', "'232'");
        this.thaiCharKey.put('é', "'233'");
        this.thaiCharKey.put('ê', "'234'");
        this.thaiCharKey.put('ë', "'235'");
        this.thaiCharKey.put('ì', "'236'");
        this.thaiCharKey.put('í', "'237'");
        this.thaiCharKey.put('î', "'238'");
        this.thaiCharKey.put('ï', "'239'");
        this.thaiCharKey.put('ð', "'240'");
        this.thaiCharKey.put('ñ', "'241'");
        this.thaiCharKey.put('ò', "'242'");
        this.thaiCharKey.put('ó', "'243'");
        this.thaiCharKey.put('ô', "'244'");
        this.thaiCharKey.put('õ', "'245'");
        this.thaiCharKey.put('ö', "'246'");
        this.thaiCharKey.put('÷', "'247'");
        this.thaiCharKey.put('ø', "'248'");
        this.thaiCharKey.put('ù', "'249'");
        this.thaiCharKey.put('ú', "'250'");
        this.thaiCharKey.put('û', "'251'");
	}

    public void Connect(SocketAddress ip, int port)
    {
        try {
			client.connect(ip, port);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
    }

    public void Close()
    {
        try {
			client.close();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
    }

    public Boolean isConnected()
    {
        return client.isConnected();
    }

    public Boolean isClosed()
    {
        return client.isClosed();
    }

    public void Send(String message)
    {
    	if (client.isConnected() && !client.isClosed()) {
	        try
	        {
	        	OutputStream output = client.getOutputStream();
	        	output.write(convertThaiCharToCode(message).getBytes());
	        	output.flush();
	        }
	        catch (Exception e)
	        {
	        }
    	}
    }

    public String Receive()
    {
        String responseMsg = "";
    	if (client.isConnected() && !client.isClosed()) {
	        try
	        {
	            InputStream input = client.getInputStream();
	            if (input.available() > 0) {
	                byte[] byteInput = new byte[input.available()];
	                int reads = input.read(byteInput, 0, byteInput.length);
	                if(reads < byteInput.length)	//Just to be sure, not necessary I think
	                {
	                    byte[] byteInput_ = new byte[reads];
	                    System.arraycopy(byteInput, 0, byteInput_, 0, reads);
	                    byteInput = byteInput_;
	                }
	                responseMsg = new String(byteInput);
	            }
	            //System.Console.WriteLine(responseMsg);
	        }
	        catch (Exception e)
	        {
	        }
    	}
        return responseMsg;
    }
    
    public InputStream getInputStream() {
    	try {
			return client.getInputStream();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			return null;
		}
    }
    
    public OutputStream getOutputStream() {
    	try {
			return client.getOutputStream();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			return null;
		}
    }
    
    private String convertThaiCharToCode(String input)
    {
        String output = input;
        for (Character key : this.thaiCharKey.keySet())
        {
            output = output.replace(String.valueOf(key), this.thaiCharKey.get(key));
        }
        return output;
    }
}

File: Client/Client/Client/Node/Network.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace MMORPGCopierClient
{
    public class Network
    {
        private TcpClient client;
        private Dictionary<Char, String> thaiCharKey;
        private Dictionary<String, Char> keyThaiChar;
        public Network(TcpClient client)
        {
            this.client = client;
            this.thaiCharKey = new Dictionary<Char, String>();
            this.keyThaiChar = new Dictionary<String, Char>();
            this.thaiCharKey.Add('ก', "'161'");
            this.thaiCharKey.Add('ข', "'162'");
            this.thaiCharKey.Add('ฃ', "'163'");
            this.thaiCharKey.Add('ค', "'164'");
            this.thaiCharKey.Add('ฅ', "'165'");
            this.thaiCharKey.Add('ฆ', "'166'");
            this.thaiCharKey.Add('ง', "'167'");
            this.thaiCharKey.Add('จ', "'168'");
            this.thaiCharKey.Add('ฉ', "'169'");
            this.thaiCharKey.Add('ช', "'170'");
            this.thaiCharKey.Add('ซ', "'171'");
            this.thaiCharKey.Add('ฌ', "'172'");
            this.thaiCharKey.Add('ญ', "'173'");
            this.thaiCharKey.Add('ฎ', "'174'");
            this.thaiCharKey.Add('ฏ', "'175'");
            this.thaiCharKey.Add('ฐ', "'176'");
            this.thaiCharKey.Add('ฑ', "'177'");
            this.thaiCharKey.Add('ฒ', "'178'");
            this.thaiCharKey.Add('ณ', "'179'");
            this.thaiCharKey.Add('ด', "'180'");
            this.thaiCharKey.Add('ต', "'181'");
            this.thaiCharKey.Add('ถ', "'182'");
            this.thaiCharKey.Add('ท', "'183'");
            this.thaiCharKey.Add('ธ', "'184'");
            this.thaiCharKey.Add('น', "'185'");
            this.thaiCharKey.Add('บ', "'186'");
            this.thaiCharKey.Add('ป', "'187'");
            this.thaiCharKey.Add('ผ', "'188'");
            this.thaiCharKey.Add('ฝ', "'189'");
            this.thaiCharKey.Add('พ', "'190'");
            this.thaiCharKey.Add('ฟ', "'191'");
            this.thaiCharKey.Add('ภ', "'192'");
            this.thaiCharKey.Add('ม', "'193'");
            this.thaiCharKey.Add('ย', "'194'");
            this.thaiCharKey.Add('ร', "'195'");
            this.thaiCharKey.Add('ฤ', "'196'");
            this.thaiCharKey.Add('ล', "'197'");
            this.thaiCharKey.Add('ฦ', "'198'");
            this.thaiCharKey.Add('ว', "'199'");
            this.thaiCharKey.Add('ศ', "'200'");
            this.thaiCharKey.Add('ษ', "'201'");
            this.thaiCharKey.Add('ส', "'202'");
            this.thaiCharKey.Add('ห', "'203'");
            this.thaiCharKey.Add('ฬ', "'204'");
            this.thaiCharKey.Add('อ', "'205'");
            this.thaiCharKey.Add('ฮ', "'206'");
            this.thaiCharKey.Add('ฯ', "'207'");
            this.thaiCharKey.Add('ะ', "'208'");
            this.thaiCharKey.Add('ั', "'209'");
            this.thaiCharKey.Add('า', "'210'");
            this.thaiCharKey.Add('ำ', "'211'");
            this.thaiCharKey.Add('ิ', "'212'");
            this.thaiCharKey.Add('ี', "'213'");
            this.thaiCharKey.Add('ึ', "'214'");
            this.thaiCharKey.Add('ื', "'215'");
            this.thaiCharKey.Add('ุ', "'216'");
            this.thaiCharKey.Add('ู', "'217'");
            this.thaiCharKey.Add('ฺ', "'218'");
            this.thaiCharKey.Add('฿', "'223'");
            this.thaiCharKey.Add('เ', "'224'");
            this.thaiCharKey.Add('แ', "'225'");
            this.thaiCharKey.Add('โ', "'226'");
            this.thaiCharKey.Add('ใ', "'227'");
            this.thaiCharKey.Add('ไ', "'228'");
            this.thaiCharKey.Add('ๅ', "'229'");
            this.thaiCharKey.Add('ๆ', "'230'");
            this.thaiCharKey.Add('็', "'231'");
            this.thaiCharKey.Add('่', "'232'");
            this.thaiCharKey.Add('้', "'233'");
            this.thaiCharKey.Add('๊', "'234'");
            this.thaiCharKey.Add('๋', "'235'");
            this.thaiCharKey.Add('์', "'236'");
            this.thaiCharKey.Add('ํ', "'237'");
            this.thaiCharKey.Add('๎', "'238'");
            this.thaiCharKey.Add('๏', "'239'");
            this.thaiCharKey.Add('๐', "'240'");
            this.thaiCharKey.Add('๑', "'241'");
            this.thaiCharKey.Add('๒', "'242'");
            this.thaiCharKey.Add('๓', "'243'");
            this.thaiCharKey.Add('๔', "'244'");
            this.thaiCharKey.Add('๕', "'245'");
            this.thaiCharKey.Add('๖', "'246'");
            this.thaiCharKey.Add('๗', "'247'");
            this.thaiCharKey.Add('๘', "'248'");
            this.thaiCharKey.Add('๙', "'249'");
            this.thaiCharKey.Add('๚', "'250'");
            this.thaiCharKey.Add('๛', "'251'");
            foreach (Char key in this.thaiCharKey.Keys)
            {
                this.keyThaiChar.Add(this.thaiCharKey[key], key);
            }
        }

        public void Connect(String ip, int port)
        {
            if (client != null && client.Connected)
                Close();
            client = null;
            client = new TcpClient();
            if (!client.Connected)
                client.Connect(ip, port);
        }

        public void Close()
        {
            if (client != null)
                client.Close();
        }

        public Boolean isConnected()
        {
            if (client == null)
                return false;
            return client.Connected;
        }

        public void Send(String message)
        {
            if (client.Connected)
            {
                Byte[] msg = System.Text.Encoding.UTF8.GetBytes(convertThaiCharToCode(message));
                NetworkStream stream = client.GetStream();
                stream.Write(msg, 0, msg.Length);
                stream.Flush();
                //Debug.WriteLine("Send: " + message);
            }
        }

        public String Receive()
        {
            String responseMsg = "";
            if (client.Connected)
            {
                Byte[] msg = new Byte[256];
                NetworkStream stream = client.GetStream();
                if (stream.DataAvailable)
                {
                    int bytes = msg.Length;
                    while (bytes >= msg.Length)
                    {
                        bytes = stream.Read(msg, 0, msg.Length);
                        responseMsg += System.Text.Encoding.UTF8.GetString(msg, 0, bytes);
                    }
                }
            }
            //Debug.WriteLine("Receive: " + responseMsg);
            return convertCodeToThaiChar(responseMsg);
        }

        private String convertThaiCharToCode(String input)
        {
            String output = input;
            foreach (Char key in this.thaiCharKey.Keys)
            {
                output = output.Replace(Convert.ToString(key), this.thaiCharKey[key]);
            }
            return output;
        }

        private String convertCodeToThaiChar(String input)
        {
            String output = input;
            foreach (String key in this.keyThaiChar.Keys)
            {
                output = output.Replace(key, Convert.ToString(this.keyThaiChar[key]));
            }
            return output;
        }
    }
}
